### PR TITLE
Move dataloader `initialize_object` to factory methods

### PR DESCRIPTION
**Pull request by hanlint**
This PR continues our effort to remove logic from `initialze_object` in preparation of deprecating yahp from the codebase. We move most of the logic into factory methods, e.g.:
```
def build_ad20k_dataloader (...) -> DataSpec
   ...

def build_synthetic_ade20k_dataloader(..) -> DataSpec
  ...
```

TODO:
- [x] ADE20k
- [x] ImageNet
- [x] C4 @growlix 
- [x] BRATS
- [X] CIFAR @dblalock 
- [x] COCO @A-Jacobson 
- [x] MNIST
- [x] LMDataset @growlix 

Need some help to complete the rest!

### CODE REVIEW

**Comment by knighton** [comment]
ready for review or wait for all checks?

**Comment by hanlint** [comment]
@knighton wait for all checks, thank you :)

**Comment by dblalock** [comment]
Added CIFAR-10 dataloader factories. Also added tests for CIFAR10 and MNIST based on `test_c4.py`. There's a bunch of duplicate code across {CIFAR,MNIST} {factory funcs, tests} right now. Might be worth separating some out into shared utilities once we have something working for all the datasets / know exactly what we'd like to DRY out.

Test code is especially redundant; only holding off on adding shared utils since maybe there's already a plan to replace `test_dataset_registry.py`.

**Comment by hanlint** [comment]
Yeah we can DRY these out later. 

@coryMosaicML since we are dropping BRATS, can you help test the imagenet/ade20k ones to make sure nothing broke with this PR? I think just kicking off a run and making sure it successfully starts training would be sufficient.

File: composer/datasets/ade20k_hparams.py
Comment (A-Jacobson) [inline composer/datasets/ade20k_hparams.py:81]:
```suggestion
            return build_synthetic_ade20k_dataloader(
                batch_size=batch_size,
                datadir=self.datadir,
                split=self.split,
                drop_last=self.drop_last,
                shuffle=self.shuffle,
                final_size=self.final_size,
                num_unique_samples=self.synthetic_num_unique_samples,
                device=self.synthetic_device,
                memory_format=self.synthetic_memory_format,
                **asdict(dataloader_hparams),
```

Comment (A-Jacobson) [reply composer/datasets/ade20k_hparams.py:81]: think you forgot to pass in datadir here and in the synthetic function?

File: composer/datasets/coco.py
Comment (A-Jacobson) [inline composer/datasets/coco.py:40]:
```suggestion
    """Builds a COCO Detection dataloader with default transforms for SSD300.
```

File: composer/datasets/coco.py
Comment (A-Jacobson) [inline composer/datasets/coco.py:53]:
```suggestion
```

File: composer/datasets/coco.py
Comment (A-Jacobson) [inline composer/datasets/coco.py:47]:
```suggestion
        shuffle (bool): whether to shuffle the dataset. Default: ``True``.
        input_size (int): the size of the input image, keep this at `300` for SSD300. Default: ``300``.
```

**Comment by hanlint** [comment]
@eracah @knighton this is ready for review!

**Review by eracah — APPROVED** [review]
Great work, @hanlint! This is super important for de-YAHPing because the dataset Hparams is where a lot of the "black magic" of hparams lives. Thanks for doing this. Looks pretty good to me, but I'll defer to James, who probably knows this stuff better than me

File: composer/datasets/lm_dataset_hparams.py
Comment (hanlint) [inline composer/datasets/lm_dataset_hparams.py:125]: @growlix can this be reworked to not use `DataloaderHparams`, but instead create the dataloader directly?

Comment (growlix) [reply composer/datasets/lm_dataset_hparams.py:126]: Fix attempted. lmk if this is what you had in mind!

**Review by A-Jacobson — APPROVED** [review]
noted what i think is a missing param in ade20k_hparams. otherwise lgtm! good stuff.

File: composer/datasets/ade20k_hparams.py
Comment (hanlint) [reply composer/datasets/ade20k_hparams.py:81]: ah, synthetic datasets don't have a datadir!

**Pull request merged**
